### PR TITLE
Add jax.scipy.linalg.companion

### PR DESCRIPTION
**Pull request by Saumay**
Adds `jax.scipy.linalg.companion` to keep chipping away at the special-matrix gap from #10144 (following `hadamard`, `circulant`, and `dft`).

The first row is `-a[1:] / a[0]` and the first sub-diagonal is ones, matching SciPy. Batching falls out of `jnp.vectorize` so `(..., N)` inputs give `(..., N-1, N-1)` outputs.

A small thing worth calling out: SciPy raises a `ValueError` when `a[..., 0] == 0`. We can't do that under jit without forcing concretization, so the function lets the `inf` / `nan` from the division propagate through. There's a `Note` in the docstring spelling this out.

### Tests
- `testCompanion`: parity vs SciPy across float / complex / int dtypes and 1-D + batched shapes (14 cases, x32 and x64).
- `testCompanionInvalidLength`: length 0 / 1 on the last axis raises.
- Doctest passes.
- Sanity-checked manually: eigenvalues of the companion of `x^3 - 6x^2 + 11x - 6` come out as `{1, 2, 3}`, and scaling all coefficients leaves the matrix unchanged.

### Notes on prior work
There were two earlier attempts that have gone quiet, and I tried to be respectful of the review history rather than start from scratch:

- #34460 covered companion specifically; @jakevdp left detailed feedback there in February that the original author didn't get back to. I read through it before writing this and tried to address each point: no separate test file, `assertRaisesRegex` for error checks, batched cases parameterized into the main test, no value-dependent error inside jit, `n < 2` matches SciPy's `ValueError`, no redundant `jnp.asarray`, and no `* 1.0` for promotion since `truediv` handles it.
- #35075 bundled five special matrices in one PR but hasn't been touched since the day it opened. This PR handles `companion` on its own; `dft` is in #37149, `fiedler` in #37334, `hadamard` is already merged (#37043), and `leslie` is in #37335.

Issue: #10144

### CODE REVIEW

File: jax/_src/scipy/linalg.py [hunk 1]
```diff
@@ -2612,6 +2612,50 @@ def _circulant(c: Array) -> Array:
...
+    A companion matrix of shape ``(..., N - 1, N - 1)``.
+
+  Note:
+    Unlike :func:`scipy.linalg.companion`, this function does not check at
+    runtime that ``a[..., 0]`` is non-zero; if the leading coefficient is
+    zero, the result will contain ``inf`` or ``nan`` entries.
+
+  Examples:
+    >>> jax.scipy.linalg.companion(jnp.array([1., -10., 31., -30.]))
+    Array([[ 10., -31.,  30.],
```
[the remark this answers is not in the document]
Comment (Saumay) [reply jax/_src/scipy/linalg.py:2684]: Applied in 9be25019. (Right to point out that `jnp.asarray` is redundant, `check_arraylike` on the prior line already rejects lists/tuples, and `jnp.atleast_1d` accepts every type that `check_arraylike` accepts.)

File: jax/_src/scipy/linalg.py [hunk 2]
```diff
@@ -2612,6 +2612,50 @@ def _circulant(c: Array) -> Array:
...
+           [  1.,   0.,   0.],
+           [  0.,   1.,   0.]], dtype=float32)
+  """
+  check_arraylike("companion", a)
```
Comment (jakevdp) [inline jax/_src/scipy/linalg.py]: Please use `a, = promote_args_inexact("companion", a)` which will call `check_arraylike` and promote the input to inexact dtype, rather than relying on the division operator below to promote implicitly.

File: tests/linalg_test.py [hunk 3]
```diff
@@ -2305,6 +2314,31 @@ def testCirculantConstruction(self, shape, dtype):
...
+     dtype=float_types + complex_types + int_types,
+  )
+  def testCompanion(self, shape, dtype):
+    rng = jtu.rand_default(self.rng())
+    def args_maker():
+      a = rng(shape, dtype)
+      lead = a[..., 0]
+      lead = np.where(lead == 0, np.ones_like(lead), lead)
+      a = a.copy()
+      a[..., 0] = lead
```
Comment (jakevdp) [inline tests/linalg_test.py]: rather than doing all this, you could use `jtu.rand_nonzero` when defining `rng`.

**Review by jakevdp — CHANGES_REQUESTED** [review]

File: jax/_src/scipy/linalg.py (hunk 2 above)
Comment (Saumay) [reply jax/_src/scipy/linalg.py]: Switched to `promote_args_inexact("companion", a)` in bae2dca4. That handles `check_arraylike` and the inexact promotion in one go, so the explicit `check_arraylike` call is gone and we no longer rely on the division to promote implicitly. Thanks for the pointer!

File: tests/linalg_test.py (hunk 3 above)
Comment (Saumay) [reply tests/linalg_test.py]: Done. `args_maker` is a one-liner now with `jtu.rand_nonzero`. Much cleaner.

**Comment by jakevdp** [comment]
There are conflicts that must be resolved here – thanks!

**Review by jakevdp — APPROVED** [review]
Thanks!
